File: ruby/rna-transcription/rna_transcription.rb
# assignment 4: rna-transcription
# author: Jason Morse
# date: May 31, 2015; updated: June 14, 2015
#

class Complement

  def self.of_dna(dna)
    unless dna.each_char.all? { |nucleotide| 'GCTA'.include?(nucleotide) }
      fail ArgumentError, "Nucleotide #{nucleotide} is not valid input (GCTA)"
    end
    dna.tr('GCTA', 'CGAU')
  end

  def self.of_rna(rna)
    unless rna.each_char.all? { |nucleotide| 'CGAU'.include?(nucleotide) }
      fail ArgumentError, "Nucleotide #{nucleotide} is not valid input (CGAU)"
    end
    rna.tr('CGAU', 'GCTA')
  end

end

File: javascript/rna-transcription/rna-transcription.js
// file: rna-transcription.js (assignment 4)
// author: Jason Morse
// date: 12/24/2015
//

var DnaTranscriber = function() {
};

var toRnaMap = {G: 'C', C: 'G', T: 'A', A: 'U'};
var toDnaMap = {C: 'G', G: 'C', A: 'T', U: 'A'};

DnaTranscriber.prototype.toRna = function(dna) {
	rna = '';
	for (var i = 0; i < dna.length; i++) {
		rna += toRnaMap[dna.charAt(i)];
	}
	return rna;
}

DnaTranscriber.prototype.toDna = function(rna) {
	dna = '';
	for (var i = 0; i < rna.length; i++) {
		dna += toDnaMap[rna.charAt(i)];
	}
	return dna;
}

module.exports = DnaTranscriber;